src/main.cc

```diff
@@ -934,13 +939,23 @@ int run_filter(StringView keystr, ConstArrayView<StringView> files, bool quiet,
             buffer_manager.delete_buffer(buffer);
         }
     }
+    catch (filter_key_error& err)
+    {
+        if (not quiet)
+            write_stderr(format("error: {}\n", err.what()));
+        goto error_cleanup;
```

Comment:
use on_scope_end to ensure clear_buffer_trash is called, and avoid goto, as it is a pretty dangerous construct in C++ (due to constructor/destructors). Here we can just return 1, which removes the need for the exit_code variable.

Reply:
We can return 1 if you don't care about freeing up memory right before the process exits, c.f. [my original reply](https://github.com/mawww/kakoune/pull/3362#issuecomment-584220469).

If you want to keep Valgrind happy I'll use `on_scope_end`, otherwise `return 1`, but I need to know which.

Reply:
Its not only valgrind, clear_buffer_trash is when buffers get effectively deleted, and run eventual hooks, when we do it has semantic meaning, not only memory freeing side effects.

Using on_scope_end ensures that gets called when we return, regardless of which return statement we use.